bank/api/api.go

```diff
@@ -0,0 +1,243 @@
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		}
+
+		cookie := http.Cookie{Name: "session", Value: session.SessionID, Expires: session.Expires}
+
+		http.SetCookie(w, &cookie)
+
+		go renew.VoidSessionAfter(sessStore, *session)
+
+		result, err := accStore.GetAccounts(session.UserID)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		}
+
+		json.NewEncoder(w).Encode(result)
+	})
+}
+
+func LogoutHandler(sessStore domain.SessionStore, renew domain.Renew) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		s := context.Get(r, "session")
+		session := s.(*domain.Session)
+
+		sessStore.DeleteSession(session.SessionID)
+
+		cookie := http.Cookie{Name: "session", Expires: time.Now()}
+
+		renew.RenewSessionCounter <- false
+
+		http.SetCookie(w, &cookie)
+		w.WriteHeader(http.StatusOK)
+	})
+}
+
+func GetUserAccounts(accStore domain.AccountStore) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		a := context.Get(r, "account")
+		account := a.(domain.Account)
+
+		result, err := accStore.GetAccounts(account.UserID)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		}
+
+		json.NewEncoder(w).Encode(result)
+	})
+}
+
+func NewUserAccount(accStore domain.AccountStore) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		a := context.Get(r, "account")
+		account := a.(domain.Account)
+
+		accountID, err := accStore.InsertAccount(account)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		err = accStore.InsertHistory(accountID, "new account", account.Currency, account.Amount)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+	})
+}
+
+func DeleteUserAccount(accStore domain.AccountStore) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		a := context.Get(r, "account")
+		account := a.(domain.Account)
+
+		result, err := accStore.GetAccount(account.UserID, account.Currency)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		err = accStore.DeleteAccount(account.UserID, account.Currency)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		err = accStore.InsertHistory(result.AccountID, "account deleted", result.Currency, 0)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+	})
+}
+
+func UserAccountWithdraw(accStore domain.AccountStore) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		a := context.Get(r, "account")
+		account := a.(domain.Account)
+
+		result, err := accStore.GetAccounts(account.UserID)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		}
+
+		for _, current := range *result {
```

Comment:
This logic should be in the persistence, as mongo already has atomic operations like: findAndUpdate. 

The following snipped
```golang
err := accountStore.Widthdraw(account,  amount)
if err != nil {
}
```
looks more reasonable to me.